### DISCUSSION TITLE
2023-07-21 - Testing #1 - The different types of testing - Chewing the Fat 📈

### DISCUSSION THREAD

**Discussion by lukecookssw** [SSWConsulting/SSW.Rules.Content#6124] [Chewing the Fat]
Hey testers!

Let's talk about [this rule](https://www.ssw.com.au/rules/different-types-of-testing/)

**Comment by leahy268** [comment]
I have seen automated testing from a distance, so it is definitely good to get a bit of an understanding.
Spent a lot of time manually testing things, though it was always very informal for me.

**Comment by brydeno** [comment]
To me the reason we test is to reduce risk of "failure". So testing effort should be commensurate with the risk.

**Reply by leehawkinsssw** [reply]
It's good to know that listening to me saying exactly this for 20 years has made it stick in your brain, Bryden!

**Comment by piers-sinclair** [comment]
Tests are critical and developers often don't have a great understanding of it. It is easy to make a mistake and introduce bugs into already functioning software.

From the Product Owners perspective it can be frustrating to see a developer introducing regressions. However, it's important to note that if you notice a pattern of regressions being implemented across multiple developers each time a problem is solved, then it may be caused by technical debt or poor testing practices. In that case, it would be wise to investigate if that is the case and what can be done about it.

There is also the team culture at play. The entire team needs to be bought in on the idea of your testing strategy and implementing it. Writing tests take a long time so if only 1 or 2 people are bought in then the rest of the team will pump out a lot more work and will look like they are doing a better job even though there is a higher chance for bugs.

**Comment by jakebayliss** [comment]
We should do another CTF series on the different kinds of tests :) Rule is a big wall of text and I think could do with some examples for each testing type

**Reply by lukecookssw** [reply]
Raise a PR :)

**Comment by MarcusIrmscherSSW** [comment]
I love the Glastonbury example in the rule 👍It shows how important load/stress testing is.

**Comment by sethdaily** [comment]
Good introduction to testing. It seems like there is lots of overlap between the types of tests, so I shall put my faith in the many people much better than me to know what to do and when!

**Comment by matt-goldman** [comment]
The rule is great! Although this comment:
> Don't confuse test approaches & techniques (focused on the "how") with types of testing (the "what"). For example, [exploratory testing](https://ssw.com.au/rules/what-is-exploratory-testing) - as an approach - applies well to several of the types of testing outlined above.

Is a huge statement for a throwaway box. This needs a rule and/or video itself!

**Reply by matt-goldman** [reply]
Also I am so bad at testing. I think it would be great if SSW ran a course on this, there's definitely a market for it and I would attend (even if I wasn't an employee).

**Comment by ozdemir-mehmet** [comment]
1 line of code, feels like 15 lines of tests.

**Comment by calumjs** [comment]
If you use GPT to generate all your tests, should you have meta-tests that ensure the generated tests are appropriate? 🤔

**Reply by Hona** [reply]
As a dev you're responsible for your usage of AI. If you get it to write a test - then break the case it is testing and it should go red. fix it & it should go green.  Blindly letting it write a test that goes green & commit is just unprofessional

**Reply by wicksipedia** [reply]
Have a look at mutation testing - its a form of testing your unit tests to make sure you don't have checks that result in false positives

**Reply by Hona** [reply]
I am aware of it - is there a nice way to integrate it into a typical dev workflow? I'm curious maybe for a 'live unit test' but a 'live mutation test'

**Reply by calumjs** [reply]
@Hona Agreed that blindly using GPT generated content is obviously not a good idea...

My point though was whether there should be another layer of the testing pyramid for meta-testing - testing your tests to make sure that when it indicates all tests are passing, that is actually meaningful.

The GPT angle was just to spark some debate, but it could apply to human-written tests just as much.

**Comment by stephanfakossw** [comment]
Nice intro into testing! Love the stress test example 😊

**Comment by SamBWagner** [comment]
I think tests fall on a spectrum of delivery <-> resilience where the more you test the more resilient your code is, but the less you deliver. It's a balance that needs to be evaluated on a case by case basis. There is no 1 size fits all solution, as with most things.

**Comment by AlexRothwell** [comment]
With the huge variety of types of tests that occur for client projects, this is a great introduction

**Comment by vladislav-kir** [comment]
I would love to see a CTF series about detailed exploration of different kinds of testing :)

**Comment by UlyssesMaclaren** [comment]
Nice explanation that I could understand :)
I'm actually interested in what the answer to that reaction button question is... is it ALL THE TESTS!!?

**Reply by PennyWalker** [reply]
Me too, I looked and thought, well you need to know it's working -> smoke test, but it has integrations, definitely 'component', but you need to know your code works... 'unit testing'... then you need to know the over all thing works, is this where 'manual/auto tests' come in... Am I now doing too many tests to get any work done? 😁

**Reply by jernejk** [reply]
It depends on how deep the client's wallets are and how much time they are willing to give for a well-tested solution. 😂
Usually, we need to go with what gives us the best value in building it reliably and the best value in keeping it working in the long run.

**Reply by leehawkinsssw** [reply]
> Am I now doing too many tests to get any work done? 

The mental switch is realizing that creating tests is just another part of the work to be done and not some separate thing.

**Comment by tkapa** [comment]
I like the idea of testing but have trouble applying the knowledge, it's an almost entirely separate art to actually making the app. Mocking is one of the biggest downsides to testing, and it's been the biggest pain point for me every single time I've had to interact with it.

I think that the rule is highly detailed, but there is some extraneous stuff that could be left out or separated.

**Comment by Jord-Gui** [comment]
Nice to learn about the many different ways of testing and what they're used for. Only heard about unit/functional/regression/systems testing before

**Comment by danielmackay** [comment]
Recently a few of us were having a discussion whether a certain type of test was a Unit Test or an Integration Test.  At the time we decided it was an Integration Test.  Turns out it was actually a Component Test!

I am aware of the pyramid and most of these testing concepts, but it was a nice refresher in this area.

**Reply by wicksipedia** [reply]
I'll die on the intergration test hill!

**Comment by bradystroud** [comment]
My experience with end-to-end tests is that they cause more pain than they solve. 👎

Other types of tests are cool, I've found AIs are pretty good at writing them.

**Comment by jeoffreyfischer** [comment]
I never heard about the different types of testing, I'm excited to learn more about each of them!

**Reply by JeanThirion** [reply]
Are you really though?

**Comment by cvionnet** [comment]
What a great idea for a CTF series!
+1 with @matt-goldman, running a session on testing would be highly valuable, especially to cover advanced topics like mocking.

**Comment by leehawkinsssw** [comment]
I'm excited to learn more about the level of understanding of these testing fundamentals amongst the SSW developer community

**Comment by AntPolkanov** [comment]
I love unit testing and try to write test if the project setup allows that. Really helps to reduce testing time and gives you confidence to change the code💪 It turns out that I mostly follow the MC⚡DC approach because it doesn't make sense to me to check only one happy path.

I never worked on a project where UI tests would justify their existence... just too many false positives😕
Has anyone had a better experience?

**Comment by Hona** [comment]
If I had only one comment that I'd hope every developer sees is to change from:

**❌ Bad Example:** Write a feature OR fix a bug, then, write a test & see it pass
**✅ Good Example (fixing a bug):** Write a test that goes red, mimicking the behaviour of the bug, then, fix the bug, then see the test pass. (if you change the test to make it pass, undo the bug fix & make sure it still fails!)
**✅ Good Example (writing a feature):** Write the feature & piece by piece write unit tests covering the branches of code that could occur. Red then green is not as important. However, it is still good to make sure a test CAN go red when it SHOULD.

TLDR; don't write false positives.

**Comment by tomek-i** [comment]
I rarely have seen projects with adequate tests.  

The companies I worked with kind of "offered" it as an add-on which of course none of the clients ever wanted due to additional cost.
Which is very ironic because almost every issue we encountered down the track, making changes, adding new features etc. could have been prevented completely or time saved investigating the cause which again would have saved heaps of time and additional cost for the client.

Anybody done or heard of **Mutation Testing** ? 
https://en.wikipedia.org/wiki/Mutation_testing

**Reply by Hona** [reply]
While a lot of us are aware and think mutation testing (e.g. Styker.NET) is cool. I haven't seen us add it to many real projects!

**Reply by wicksipedia** [reply]
Nick Chapsas' intro to testing course has a good section on mutation testing and how to integrate it into dev workflow

**Reply by tomek-i** [reply]
I haven't used it yet but on paper it does sound it would help figuring out and covering additional test scenarios.

**Reply by Hona** [reply]
Thanks for the tip Matt, I'll have to do his course now for sure

**Comment by Abf09** [comment]
Its good to learn about all these tests and their use case. This makes me more curious about what are going to be the most common tests dealing with SSW clients? and How confident we are after running these tests?

**Comment by jackreimers** [comment]
A great topic, testing is definitely a weaker spot for me so I'd love to explore it more.

**Comment by jimmy1992abc** [comment]
In the near future, AI would take over those testing, generating random data and testing the results with expectation.

**Comment by amankumarrr** [comment]
I like the video, but I found it hard to distinguish between Component testing and Integration testing based on the provided information. It seems that Component testing focuses on the logical level, while Integration testing involves communication between various component levels.

**Comment by JeanThirion** [comment]
Stress/load test is almost systematically ignored/deferred on projects, and from experience that's a very common point of failure... I wish I was more disciplined on this

**Comment by KahaMason** [comment]
A very interesting topic. It helped to shed some light on the differences between Unit tests, Integration tests and Component tests which I now understand the distinct differences. The Pyramid was a great visual description on the levels of tests.

**Comment by wicksipedia** [comment]
Personally, I think the testing pyramid is overused - it sets an expectation that you need to write lots of unit tests. Whereas the types of tests you need is really determined by what you are building, how it is used, and what budget the client has. At the end of the day, it's all attempts at risk mitigation.

My AI Council summed up the page for me:

> So, this SSW.Rules webpage is basically a goldmine for anyone wanting to get the lowdown on software testing. It's like the "Netflix of Testing" – it's got everything from the classics like unit testing and regression testing to the indie hits like consumer-driven contract testing.

**Reply by leehawkinsssw** [reply]
"At the end of the day, it's all attempts at risk mitigation." Music to my ears. The point of different types of tests are to mitigate different types of risk.

**Comment by Ben0189** [comment]
Testing is important in software development, but doing all test above in great details can be time-consuming and resource-intensive. Understand the value of each test given before doing the test is also important so that developer can achieving a better balance between comprehensive testing and efficient testing practices

**Comment by Khaled-Albahsh** [comment]
Automated testing is cool, but not made for everything. Sometimes it can be a healthy mix to use it alongside manual testing. It helps provide some clarity on things that automated test might not give. ( Watch things break in front of you LIVE )

**Comment by JackDevAU** [comment]
I think testing is super important, however I find it hard to add to a lot of the projects I've worked on as most of the products we create are MVP's where feature development is often the focus. I think as others have stated a CtF series on testing would be super beneficial!

**Comment by landonmaxwell** [comment]
Fun to see the differences between testing 👍

**Comment by MattParkerDev** [comment]
I love this exploration into testing. Unit tests are minimum effort to ensure your code does what you expect. Integration tests should not be overused, where budget allows

**Comment by Harry-Ross** [comment]
Thought the rule was a great introduction to testing - a great resource for SSW devs needing quality definitions on the types of testing. I would love to learn more about the actual implementation of these tests in code

**Comment by AshAnil** [comment]
Nice rule ⭐ It is good to know we had different types of testing.

**Comment by chrisschultzssw** [comment]
Interesting stuff. I didn't know much about testing before, so this was a good intro.

**Comment by bahjatalaadelssw** [comment]
The pyramid diagram helped explain the priorities and preferences for testing.

**Comment by steven0x51** [comment]
Great rule, I'm keen to learn some detailed testing implementation for different types of testing

**Comment by PennyWalker** [comment]
I am really going to enjoy looking at the feedback from this one. I find testing really interesting because they are necessary, but they also increase the cost of a project, so getting the balance right is critical. I'm interested in how you strike that balance. Loved the video, it was really easy to follow and understand for me!

**Comment by CoffeeBugHub** [comment]
Good introduction of different tests. Would keen to know more in detail in terms of implementations in different scenarios.

**Comment by jomelsontanco** [comment]
I started my career as a tester then instead of delving deeper technically (a.k.a become a software engineer), I did more functional work (a.k.a. functional/business analyst). 

Along the way I got acquainted with the ISO standard for software testing ISO/IEC 9126, as well as got acquainted with the Australian author of this said standard (message me privately if you are keen to know more).

A couple of (opinionated) observations from experience:
- The pyramid is a lie in practice.  It's more of an obese michelin mascot. There's a lot of factors to this, number one is no one really understands the business requirements at the beginning of a big project.
- Automation is hard. Outsourcing to offshore delivery centers seems to be the game.
- End to End tests are out of the expertise of most software developers.
- Unit tests are dead because of the heavy use of COTS (commercial-of-the-shelf) software by businesses.

TLDR; Testing relies a lot on a good intution of a specific domain. This expertise is usually non-transferrable. It is functional first, technical second. 

With this CTF, this is what I want to improve for myself:
- For new .NET/web developers like myself, I would probably stick with honing my craft. This means writing the best unit tests as much as I can, do as much component and integrating testing that I can.
- Leave other forms of testing to functional experts.

**Comment by JackDuan123** [comment]
Love today's video.
But not a big fan of 100% code coverage, or using code coverage as indicator of how good the unit tests are done.
This is called goal dis-placement, we have better number to evaluate the quality of unit test and testcoverage.

**Comment by jernejk** [comment]
I have never considered smoke tests to be automatable. 😯

It's great but I think it would be great if we had a bit more explanations on some testing types.
As mentioned, "Smoke testing" usually is within the context of manual testing. Where does component testing fit in, and do all of the types of testing have automated and manual variants or only some?

I generally prefer component testing over unit (provided I can manage dependencies easily) as it tells me if the overall thing is working right but I do write a lot of unit tests myself.
For projects like TimePro, I write quite a bit of black box integration tests, partly due to migration to .NET 8 but also because it's an older application with many strings attached and many things that could go wrong. It makes sleep at night a bit easier, knowing that integration tests are passing. 😂

For coverage, 100% on the critical parts that we can't afford to be broken. (like saving invoices and timesheets!) The rest is nice to have. 😊

**Comment by christoment** [comment]
Really like the rule and guidelines in the rule, especially the table https://www.ssw.com.au/rules/different-types-of-testing/#8-types-of-testing :)

Will definitely keep this handy for future client projects 👍

**Comment by jerryluossw** [comment]
Great rule, and I think we really need a series of CTF to learn each type of testing

**Comment by babakamyljanovssw** [comment]
This rule is great!!! I had an experience using regression testing in my last internship, and it was very complicated and time-consuming. This rule gave me brief explanation about other test types, and its usage. Looking forward to seeing another CTH about this topic!

**Comment by jimmidier** [comment]
This is to me, another perspective of saying, don't reinvent the wheel. Not only because wheels reduce development, but as wheels are well tested and version controlled by various vendors, they reduce testing efforts on the end application as well. 

Another way of thinking it, on certain situations it could also mean DO invent the wheel. Extracting, testing and packaging different levels of infrastructures reduces the stress on one team, and in a certain way, mitigates the risks by separating concerns to different projects.

**Comment by jontygardner** [comment]
Testing is so important so the end-user has a great experience! The last point really hit the nail on the head about manual testing. It is much better to break down the type of tests into different stages so you can quickly identify the bug. 

It would be awesome if we could do something like this for the TV Team. An AI could watch our videos and look for typos, and then another AI tests for any glitches. Right now, it is too subjective that you need human eyes and ears to test please every video.

**Comment by Marxoz** [comment]
I really think it is worth diving deeper into testing, testing techniques, and when to use what. Not only does testing pick up issues before the solution goes into production, but it also simplifies the work for the developer to fix the issue as it narrows down where the issue is happening. Can't wait to hear more.

**Comment by sylhuang** [comment]
I often find it more time-consuming to write appropriate unit tests/e2e tests than writing the actual feature or fixing the bug. Some tests could be extremely difficult to write especially when it is state-dependent and is part of a complex business logic. Other time, line coverage is not as critical or accurate (e.g. testing on each line of css property while we just need to make sure UI looks a certain way).

**Comment by Aibono1225** [comment]
Love the rule because it just gave me more insights about testing. I often have the feeling that implementing tests took too much time, but I also understand its importance in all cases. As we said, it's crucial to discover any bugs at the beginning rather than later on.

**Comment by johnxu16** [comment]
Spending much time writing unit tests to reach a high percentage of code coverage is unrealistic sometimes.
1. Not all functions worth writing a unit test
2. Devs cannot think of many cases thoroughly as testers do

Based on my working experience, the smoking test is extremely important since it ensures every deployment is functional. E2E is hard to implement and only be used on the mobile end. Regression-test will be required much more often than your expectation since the complexity of projects grows fast nowadays.

**Comment by tino-liu** [comment]
Just yesterday I wrote an unit test that covers a method for checking data payload that covers all the bad data I could think of. What it should do is expect an argument exception and pass the test. However, the test failed and the bad data went through. I found out later that there were cases I didn't initially thought of when developing the feature. So now I have a bug to fix. Just maybe this could be avoided if it was done in TDD style.
Testing in action!

**Comment by KristenHu** [comment]
I like the video, I didn't know testing well before, but it makes me know it more systematically.

**Comment by tombui99** [comment]
Testing is highly crucial in software engineering yet only few clients I have worked with spend time to write tests because of budget. With testing, we can easily identify bugs, ensure software quality, validate requirements and so much more. It guarantees a reliable and secure product, meeting user expectations and industry standards

**Reply by leehawkinsssw** [reply]
Just a bit of advice, Tom: use safer language than "guarantees" when discussing the benefits of testing with clients. No amount of testing - of any type - can guarantee a bug-free product https://www.ssw.com.au/rules/complete-testing-is-impossible/ Making stronger claims about your testing than you can possibly back up immediately puts you in the prime spot for blame if issues are found in production.

**Comment by zl24622** [comment]
Nice video. The first time heard about the Testing Pyramid, a few other knowledge learned from the rule and video.

I remember reading a blog in which the author says he is not doing Test-Driven-Development, instead, he is doing Testbility-Driven-Development. If he notices he has to write many tests or do a lot of mocks, probably means his class or function or even the whole architecture is too complicated or has too much dependence on each other. He would spend more time refactoring his code.

I feel his approach is more realistic because, in the real world, it's always a balance between time, and $ and quality. Most of the time we probably cannot write all tests, but can use tests to help keep a mindset for developing simple, clean applications.

**Comment by PothieuG** [comment]
Great video and nice summary in the rule !
I never hearded about Testing Pyramid/
I'm not a testing expert and it's still difficult for me to appreciate how far I should test the code or not.
I really want to see and learn all the testing strategies I saw today.

**Comment by drwharris** [comment]
I find people either 

1. Ignore testing all together, 
2. Do the bare minimum just to ✔️ a box
3. Go overboard and over complicate things leading to unreliable testing

People often forget the importance of keeping tests simple and isolated.

**Comment by william-liebenberg** [comment]
Looking forward to seeing when most of us write tests. Before, during, or after building a feature.

Are we a group of TDD purists (before) or do we think a bit differently and write code towards the user requirement first, then write tests to cover what we have done.?

**Comment by BrookJeynes** [comment]
I think testing is important but it's heavily dependent on what type of work is being done. Sometimes tests are too difficult to mock or are just simply not needed. Other times, having a strong suite of unit tests can ensure that over time as the application changes, the intended functionality remains the same.

**Comment by AttackOnMorty** [comment]
Unit test is essential for refactoring. It is a minimal guarantee that the behaviour has not been modified.

**Comment by Levijj22** [comment]
Great video! Simple explanation of what seems to be a vast and complicated topic.
The pyramid illustrates it well.

**Comment by Freego1783** [comment]
Nice to see a CTF on Testing, I know I lack some skills there.

I have often encountered clients who explicitly requested not to write any tests because they didn't want the devs to spend time on them and wanted the devs to focus on the features. Those client considered it was a waste of money, and obviously, they were impossible to convince.

**Comment by TobyChurches** [comment]
Great CTF! I love the breakdown of the different types of testing! This better summarised the information from a entire course at University 😁

**Comment by ozair-ashfaque** [comment]
Awesome CTF! I would like to go deeper in it
